### ISSUE TITLE
Pyright does not detect correct type of TypedDict when iterating over .items()

### ISSUE THREAD

**Issue by abirnkraut-tkb** [microsoft/pyright#7407]
I have created a TypedDict with different types. When iterating over the items of the typed dict and switching cases with if statements pyright does not detect the correct type. Example in the following code snippet.

```
from typing import TypedDict

class MyObject(TypedDict):

    key1: str
    key2: int

myobject: MyObject = {
    "key1": "My first string",
    "key2": 5
}

for key, value in myobject.items():
    if key == "key1":
        # assert isinstance(value, str)
        output = value.split()
    if key == "key2":
        # assert isinstance(value, int)
        output = [value/2]
```

Currently I'm getting errors such as 'Cannot access member "split" for type "object"' or 'Operator "/" not supported for types "object" and "Literal[2]". I can only avoid these errors when explicitly checking the type by something like `assert isinstance(value, str)` or `assert isinstance(value, int)`.

I'd like pyright to detect the correct type, based on the if-filters on the keys in the example above.

**Comment by erictraut** [comment]
This isn't how type narrowing works. The type of `key` is `str` and the type of `value` is `object`. By the time the nested `if` statements are encountered, there is no relationship between `key` and `value`. They are completely independent variables. Type checkers do not track "entanglement" between independent variables.

I recommend changing your code structure to directly access keys from the dictionary object. This approach will work with the static type checker, not against it.

If you want to stick with your current code structure, you'll need to use `assert` statements.
